docs/src/main/asciidoc/using-vertx.adoc

```diff
@@ -0,0 +1,483 @@
+
+To asynchronously greet a client, the endpoint method must return a `java.util.concurrent.CompletionStage`:
+
+[source,java]
+----
+@Path("/hello")
+public class GreetingResource {
+
+    @Inject
+    Vertx vertx;
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    @Path("{name}")
+    public CompletionStage<String> greeting(@PathParam("name") String name) {
+        // When complete, return the content to the client
+        CompletableFuture<String> future = new CompletableFuture<>();
+
+        long start = System.nanoTime();
+
+        // TODO: asynchronous greeting
+
+        return future;
+    }
+}
+----
+
+So far so good.
+Now let's use the Vert.x API to implement the artificially delayed reply with the `setTimer` provided by Vert.x:
+
+[source,java]
+----
+// Delay reply by 10ms
+vertx.setTimer(10, l -> {
+    // Compute elapsed time in milliseconds
+    long duration = MILLISECONDS.convert(System.nanoTime() - start, NANOSECONDS);
+
+    // Format message
+    String message = String.format("Hello %s! (%d ms)%n", name, duration);
+
+    // Complete
+    future.complete(message);
+});
+----
+
+That's it.
+Now start {project-name} in `dev` mode with:
+
+[source, shell]
+----
+mvn compile quarkus:dev
+----
+
+Eventually, open your browser and navigate to http://localhost:8080/hello/Quarkus, you should see:
+
+[source, shell]
+----
+Hello Quarkus! (10 ms)
+----
+
+=== Streaming using Server-Sent Events
+
+{project-name} web resources that need to send content as https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events[server-sent events] must have a method:
+
+* declaring the `text/event-stream` response content type
+* returning a https://www.reactive-streams.org/[Reactive Streams] `Publisher`
+
+In practice, a streaming greeting service would look like:
+
+[source,java]
+----
+@Path("/hello")
+public class StreamingResource {
+
+    @GET
+    @Produces(MediaType.SERVER_SENT_EVENTS)
+    @Path("{name}/streaming")
+    public Publisher<String> greeting(@PathParam("name") String name) {
+        // TODO: create a Reactive Streams publisher
+        return publisher;
+    }
+}
+----
+
+How to create a Reactive Streams publisher?
+There are a few ways to do this:
+
+1. If you use `io.vertx.axle.core.Vertx`, the API provides `toPublisher` methods (and then use RX Java 2 or Reactive Streams Operators to manipulate the stream)
+2. You can also use `io.vertx.reactivex.core.Vertx` which already provides RX Java 2 (RX Java 2 `Flowable` implement Reactive Streams `publisher`).
+
+The first approach can be implemented as follows:
+
+[source, java]
+----
+// Use io.vertx.axle.core.Vertx;
+@Inject Vertx vertx;
+
+@GET
+@Produces(MediaType.SERVER_SENT_EVENTS)
+@Path("{name}/streaming")
+public Publisher<String> greeting(@PathParam("name") String name) {
+    return ReactiveStreams.fromPublisher(vertx.periodicStream(2000).toPublisher())
+            .map(l -> String.format("Hello %s! (%s)%n", name, new Date()))
+            .buildRs();
```

Comment:
@cescoffier I hope you have a good reason to have named this method `buildRs()` and not `toPublisher()` because from a user PoV I just did `toPublisher`, `fromPublisher` and then `buildRs` sticks out like a bloody thorn, especially when I want to return a damn `publisher`

Reply:
This comes from the spec. In Java 9 it will be `build` (and use Flow).
I hope it would be possible to return a `PublisherBuilder` with the right _provider_.

Reply:
Well, the spec was crazy to allow a method name with an abbreviation like `buildRs`, sorry :(

Reply:
It's the JDK8/JDK9 issue... because we had to plan backward compatibility we cannot use `build()`:

```
/*ReactiveStream*/ Publisher<X> build();
/*Flow*/ Publisher<X> build();
```

While I agree it's not great, it's going to be hidden as soon as Flow becomes the main model.

Reply:
Well, `buildReactiveStream()` was still available ;)